File: experiments/issue-288/ground_truth.mjs
import { Parser } from '../../js/src/Parser.js';
const parser = new Parser();
const render = (n) =>
  !n.values || n.values.length === 0
    ? `<${n.id ?? ''}>`
    : `(${n.id !== null && n.id !== undefined ? `<${n.id}>: ` : ''}${n.values.map(render).join(' ')})`;
const cases = [
  '(a " " b)', '(a "" b)', "(a '' b)", '(a `` b)',
  '(a "" "" b)', "(a '' '' b)", '(a `` `` b)',
  '(a ""x"" b)', '(a """" b)', '(x "" " "")', `(x ' " ')`,
  '("" ("" 1))', `("" ('' 1))`, '("x" ("" 1))', '("" ("x" 1))',
  '("" x ("" 1))', '("" 1 ("" 1))', '(o: ("" (o: ("" 1))))',
  '(a " b)', '(a """ b)', '("")', '("": 1)', '(a ""  "" b)', '("" "")',
];
for (const c of cases) {
  try {
    const r = parser.parse(c);
    console.log(c.padEnd(24), '=>', r.map(render).join('\n'));
  } catch (e) {
    console.log(c.padEnd(24), '=> Err(' + e.message.split('\n')[0] + ')');
  }
}

File: experiments/issue-288/ground_truth.rs
//! Prints the canonical rendering of every case in issue #288, so the
//! conformance table can be checked against a real parse in every language.
use links_notation::{parse_lino_to_links, LiNo};

fn render(node: &LiNo<String>) -> String {
    match node {
        LiNo::Ref(id) => format!("<{}>", id),
        LiNo::Link { id, values } => {
            let head = id
                .as_ref()
                .map(|id| format!("<{}>: ", id))
                .unwrap_or_default();
            let body = values.iter().map(render).collect::<Vec<_>>().join(" ");
            format!("({}{})", head, body)
        }
    }
}

fn main() {
    let cases: Vec<&str> = vec![
        r#"(a " " b)"#,
        r#"(a "" b)"#,
        r#"(a '' b)"#,
        "(a `` b)",
        r#"(a "" "" b)"#,
        r#"(a '' '' b)"#,
        "(a `` `` b)",
        r#"(a ""x"" b)"#,
        r#"(a """" b)"#,
        r#"(x "" " "")"#,
        r#"(x ' " ')"#,
        r#"("" ("" 1))"#,
        r#"("" ('' 1))"#,
        r#"("x" ("" 1))"#,
        r#"("" ("x" 1))"#,
        r#"("" x ("" 1))"#,
        r#"("" 1 ("" 1))"#,
        r#"(o: ("" (o: ("" 1))))"#,
        r#"(a " b)"#,
        r#"(a """ b)"#,
        r#"("")"#,
        r#"("": 1)"#,
        r#"(a ""  "" b)"#,
        r#"("" "")"#,
    ];
    for case in cases {
        match parse_lino_to_links(case) {
            Ok(links) => {
                let rendered = links.iter().map(render).collect::<Vec<_>>().join("\n");
                println!("{:<24} => {}", case, rendered);
            }
            Err(e) => println!("{:<24} => Err({})", case, e),
        }
    }
}
